pkg/template/processor.go

```diff
@@ -0,0 +1,111 @@
+package template
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	"context"
+	"github.com/go-logr/logr"
+	templatev1 "github.com/openshift/api/template/v1"
+	"github.com/openshift/library-go/pkg/template/generator"
+	"github.com/openshift/library-go/pkg/template/templateprocessing"
+	errs "github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Processor struct {
+	cl     client.Client
+	scheme *runtime.Scheme
+	log    logr.Logger
+}
+
+func NewProcessor(cl client.Client, scheme *runtime.Scheme, log logr.Logger) *Processor {
+	return &Processor{cl: cl, scheme: scheme, log: log}
+}
+
+func (p Processor) ProcessAndApply(tmplContent []byte, values map[string]string) error {
+	objs, err := p.Process(tmplContent, values)
+	if err != nil {
+		return err
+	}
+
+	return apply(p.cl, objs, p.log)
+}
+
+func apply(cl client.Client, objs []runtime.RawExtension, rq logr.Logger) error {
+	for _, obj := range objs {
+		if obj.Object == nil {
+			rq.Info("template object is nil")
+			continue
+		}
+		gvk := obj.Object.GetObjectKind().GroupVersionKind()
+		rq.Info("processing object", "version", gvk.Version, "kind", gvk.Kind)
+		if err := cl.Create(context.TODO(), obj.Object); err != nil {
+			if errors.IsAlreadyExists(err) {
+				// If client failed to create all resources(few created, few remaining) in the first run, then to avoid
+				// resource already existing errors for later runs.(it's like oc apply, don't fail to exists)
+				continue
+			}
+			rq.Error(err, "unable to create resource", "type", gvk.Kind)
+			return errs.Wrapf(err, "unable to create resource of type %s", gvk.Kind)
+		}
+	}
+	return nil
+}
+
+// Process processes the template with the given content
+func (p Processor) Process(tmplContent []byte, values map[string]string) ([]runtime.RawExtension, error) {
+	scheme := p.scheme
+	codecs := serializer.NewCodecFactory(scheme)
```

Comment:
I checked, it seems that creation of `CodecFactory` does lot of processing so it's better to add this as field to `Processor` struct and init only once.